src/utility.jl

```diff
@@ -33,25 +27,58 @@ function taylorexpansion(graph::G, var_dependence::Dict{Int,Vector{Bool}}=Dict{I
             coeff = Graph([]; operator=Sum(), factor=graph.factor)
             result.coeffs[o] = coeff
         end
+        taylormap[graph.id] = result
         return result
     else
-        taylormap = Dict{Int,TaylorSeries{G}}() #Saves the taylor series corresponding to each nodes of the graph
-        for g in Leaves(graph)
-            if !haskey(taylormap, g.id)
-                taylormap[g.id] = taylorexpansion(g, var_dependence)
-            end
-        end
+        taylormap[graph.id] = apply(graph.operator, [_taylorexpansion!(sub, var_dependence; taylormap=taylormap) for sub in graph.subgraphs], graph.subgraph_factors)
+        return taylormap[graph.id]
+    end
+end
 
-        rootid = -1
-        for g in PostOrderDFS(graph) # postorder traversal will visit all subdiagrams of a diagram first
-            rootid = g.id
-            if isleaf(g) || haskey(taylormap, g.id)
-                continue
-            end
-            taylormap[g.id] = apply(g.operator, [taylormap[sub.id] for sub in g.subgraphs], g.subgraph_factors)
-        end
-        return taylormap[rootid]
+
+"""
+    function taylorexpansion!(graph::G, taylormap::Dict{Int,TaylorSeries{G}}(), var_dependence::Dict{Int,Vector{Bool}}=Dict{Int,Vector{Bool}}()) where {G<:Graph}
+
+    Return a taylor series of graph g. If variable dependence is not specified, by default, assume all leaves of graph depend on all variables. The taylor series of all nodes of graph is 
+    saved in the taylormap dictionary.
+
+#Arguments
+
+- `graph` Target graph.
+- `var_dependence::Dict{Int,Vector{Bool}}` The variables graph leaves depend on. Should map each leaf ID of g to a Vector{Bool},
+    indicating the taylor variables it depends on. By default, assumes all leaves depend on all variables.
+- `taylormap::Dict{Int,TaylorSeries{G}}` The taylor series correponding to graph nodes. Should map each graph node ID (does not need to belong to input graph) to a taylor series.
+    All new taylor series generated by taylor expansion will be added into the expansion.
+"""
+function taylorexpansion!(graph::G, var_dependence::Dict{Int,Vector{Bool}}=Dict{Int,Vector{Bool}}(); taylormap::Dict{Int,TaylorSeries{G}}=Dict{Int,TaylorSeries{G}}()) where {G<:Graph}
```

Comment:
We can probably refactor these `taylorexpansion!` and `taylorexpansion` methods such that the intermediate function is `_taylorexpansion` is not necessary. The structure should be similar to the transformations defined for a single graph & vector of graphs in optimize.jl.